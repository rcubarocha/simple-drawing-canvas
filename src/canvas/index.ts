import { RGBColor } from '../color';

interface ICoords {
    x: number,
    y: number,
}

type TouchEventsMap = Pick<HTMLElementEventMap, 'touchstart' | 'touchmove' | 'touchend' | 'touchcancel'>;

export type CanvasToolType = 'pen' | 'bucket' | 'eraser' | 'clear';

type CanvasAction = {
    tool: CanvasToolType,
    start?: ICoords,
    end?: ICoords,
    size?: number,
    color?: string,
}

interface IDrawingCanvas {
    canvasElement: HTMLCanvasElement,
    width: number,
    height: number,
    scale: number,
    toolType: CanvasToolType
    brushSize: number,
    brushColor: RGBColor,
    lastCoords: ICoords,
    toolDown: boolean,
    actionsHistory: CanvasAction[],
    historyIndex: number,
    initCanvas(): void,
}

type MouseEventCallback = (e: MouseEvent) => void;

export class DrawingCanvasController implements IDrawingCanvas {
    canvasElement: HTMLCanvasElement;

    width: number;

    height: number;

    scale: number = 1;

    toolType: CanvasToolType = 'pen';

    brushSize: number = 5;

    brushColor: RGBColor = new RGBColor(0, 0, 0);

    lastCoords: ICoords = {
      x: 0,
      y: 0,
    };

    toolDown: boolean = false;

    actionsHistory: CanvasAction[] = [];

    historyIndex: number = 0;

    // boundMouseCallback = this.onCanvasEvent.bind(this);
    // boundTouchCallback = this.onTouchEvent.bind(this);

    canvasFocusCallback = (e: TouchEvent) => {
      if (e.target === this.canvasElement) {
        e.preventDefault();
      }
    }

    static eventMap = new Map<keyof TouchEventsMap, keyof HTMLElementEventMap>([
      ['touchstart', 'mousedown'],
      ['touchmove', 'mousemove'],
      ['touchend', 'mouseup'],
      ['touchcancel', 'mouseup'],
    ]);

    toolFuncs = new Map<CanvasToolType, MouseEventCallback>([
      ['pen', this.penTool.bind(this)],
      ['clear', this.clearTool.bind(this)],
      ['bucket', this.bucketTool.bind(this)],
      ['eraser', this.eraserTool.bind(this)],
    ]);

    constructor(canvasElement: HTMLCanvasElement, width: number, height: number) {
      this.canvasElement = canvasElement;
      this.width = width;
      this.height = height;

      this.initCanvas();
    }

    // eslint-disable-next-line class-methods-use-this
    drawLine(ctx: CanvasRenderingContext2D, from: ICoords, to: ICoords, width: number, color: string) {
      ctx.beginPath();
      ctx.lineCap = 'round';
      ctx.lineJoin = 'round';
      ctx.lineWidth = width;
      ctx.moveTo(from.x, from.y);
      ctx.lineTo(to.x, to.y);
      ctx.strokeStyle = color;
      ctx.stroke();
      ctx.closePath();
    }

    performCanvasAction(action: CanvasAction) {
      const ctx = this.canvasElement.getContext('2d')!;

      switch (action.tool) {
        case 'pen':
          ctx.globalCompositeOperation = 'source-over';

          this.drawLine(ctx, action.start!, action.end!, action.size!, action.color!);
          break;
        case 'clear':

          ctx.globalCompositeOperation = 'source-over';

          ctx.clearRect(0, 0, this.canvasElement.width, this.canvasElement.height);
          break;
        case 'bucket':
          ctx.globalCompositeOperation = 'source-over';

          ctx.fillStyle = action.color!;
          ctx.fillRect(0, 0, this.canvasElement.width, this.canvasElement.height);
          break;
        case 'eraser':
          ctx.globalCompositeOperation = 'destination-out';

          this.drawLine(ctx, action.start!, action.end!, action.size!, '#ffffff');
          break;
        default:
          break;
      }
    }

    penTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
        const { x, y } = this.canvasElement.getBoundingClientRect();
        this.lastCoords.x = (e.clientX - x) * this.scale;
        this.lastCoords.y = (e.clientY - y) * this.scale;
      } else if (e.type === 'mousemove') {
        if (!this.toolDown) return;

        const { x, y } = this.canvasElement.getBoundingClientRect();

        const newCoords: ICoords = {
          x: (e.clientX - x) * this.scale,
          y: (e.clientY - y) * this.scale,
        };

        const action: CanvasAction = {
          tool: 'pen',
          start: { ...this.lastCoords },
          end: { ...newCoords },
          size: this.brushSize,
          color: this.brushColor.toHex(),
        };

        this.actionsHistory.push(action);

        this.performCanvasAction(action);

        this.lastCoords = newCoords;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        if (e.target === this.canvasElement) {
          const { x, y } = this.canvasElement.getBoundingClientRect();

          const newCoords: ICoords = {
            x: (e.clientX - x) * this.scale,
            y: (e.clientY - y) * this.scale,
          };

          const action: CanvasAction = {
            tool: 'pen',
            start: { ...this.lastCoords },
            end: { ...newCoords },
            size: this.brushSize,
            color: this.brushColor.toHex(),
          };

          this.actionsHistory.push(action);

          this.performCanvasAction(action);

          this.lastCoords = newCoords;
        }

        this.toolDown = false;
      }
    }

    clearTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        // Only clear the canvas if the click release happens inside the canvas
        if (e.target === this.canvasElement) {
          const action: CanvasAction = { tool: 'clear' };

          this.actionsHistory.push(action);

          this.performCanvasAction(action);
        }

        this.toolDown = false;
      }
    }

    bucketTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        // Only clear the canvas if the click release happens inside the canvas
        if (e.target === this.canvasElement) {
          // const action: CanvasAction = { tool: 'bucket', color: this.brushColor.toHex() };

          // this.actionsHistory.push(action);

          // this.performCanvasAction(action);

          // this.playDrawing();
          this.performAllCanvasActions();
        }

        this.toolDown = false;
      }
    }

    eraserTool(e: MouseEvent) {
      if (e.type === 'mousedown') {
        this.toolDown = true;
        const { x, y } = this.canvasElement.getBoundingClientRect();
        this.lastCoords.x = (e.clientX - x) * this.scale;
        this.lastCoords.y = (e.clientY - y) * this.scale;
      } else if (e.type === 'mousemove') {
        if (!this.toolDown) return;

        const { x, y } = this.canvasElement.getBoundingClientRect();

        const newCoords = {
          x: (e.clientX - x) * this.scale,
          y: (e.clientY - y) * this.scale,
        };

        const action: CanvasAction = {
          tool: 'eraser', start: { ...this.lastCoords }, end: { ...newCoords }, size: this.brushSize,
        };

        this.actionsHistory.push(action);

        this.performCanvasAction(action);

        this.lastCoords = newCoords;
      } else if (e.type === 'mouseup') {
        if (!this.toolDown) return;

        if (e.target === this.canvasElement) {
          const { x, y } = this.canvasElement.getBoundingClientRect();

          const newCoords = {
            x: (e.clientX - x) * this.scale,
            y: (e.clientY - y) * this.scale,
          };

          const action: CanvasAction = {
            tool: 'eraser', start: { ...this.lastCoords }, end: { ...newCoords }, size: this.brushSize,
          };

          this.actionsHistory.push(action);

          this.performCanvasAction(action);

          this.lastCoords = newCoords;
        }

        this.toolDown = false;
      }
    }

    performAllCanvasActions() {
      const action: CanvasAction = { tool: 'clear' };

      this.performCanvasAction(action);

      this.actionsHistory.pop();

      this.actionsHistory.forEach((ca) => {
        this.performCanvasAction(ca);
      });
    }

    playDrawing() {
      const action: CanvasAction = { tool: 'clear' };

      this.performCanvasAction(action);

      this.historyIndex = 0;

      const interval = setInterval(() => {
        if (this.actionsHistory.length > 0) this.performCanvasAction(this.actionsHistory[this.historyIndex]!);

        if (this.historyIndex < this.actionsHistory.length - 1) {
          this.historyIndex += 1;
        } else {
          clearInterval(interval);
        }
      }, 5);
    }

    onCanvasEvent = (e: MouseEvent) => {
        // console.log('-- On Canvas Event --');
        // console.log({ type: e.type, target: e.target, current: e.currentTarget });
        this.toolFuncs.get(this.toolType)!(e);
    }

    onTouchEvent = (e: TouchEvent) => {
      // console.log('-- On Touch Event --');
      // console.log({ type: e.type, target: e.target, current: e.currentTarget });

      const me = new MouseEvent(DrawingCanvasController.eventMap.get(e.type as keyof TouchEventsMap)!, {
        clientX: e.changedTouches[0].clientX,
        clientY: e.changedTouches[0].clientY,
        bubbles: true,
      });

      /**
       * TouchEnd always targets the target of the preceding TouchStart event
       * even if the pointer has moved away from that element
       * We get the element over which the TouchEnd happened using elementFromPoint()
       * so we can accurately trigger the synthetic mouse event
       */
      const doc = this.canvasElement.ownerDocument!;
      const touchEndElement = doc.elementFromPoint(me.clientX, me.clientY)!;

      if (doc && touchEndElement) {
        touchEndElement.dispatchEvent(me);
      }

      // if (e.type === 'touchend' || e.type === 'touchcancel') {
      //     this.canvasElement.ownerDocument!.dispatchEvent(me);
      //     console.log('Dispathcing MouseUP from TouchEnd/Cancel')
      //     // this.canvasElement.dispatchEvent(me);
      // }
      // else {
      //     this.canvasElement.dispatchEvent(me);
      // }
    }

    onWindowResize = () => {
      this.scale = this.width / this.canvasElement.offsetWidth;

      this.canvasElement.style.height = `${this.height / this.scale}px`;
    }

    reset() {
      this.actionsHistory = [];
      this.historyIndex = 0;
      this.lastCoords = {
        x: 0,
        y: 0,
      };
      this.toolDown = false;
    }

    teardown() {
      DrawingCanvasController.eventMap.forEach((_, k) => {
        if (k === 'touchend' || k === 'touchcancel') {
                this.canvasElement.ownerDocument!.removeEventListener(k, this.onTouchEvent);
        } else {
          this.canvasElement.removeEventListener(k, this.onTouchEvent);
        }

            // Prevent scrolling on touch interfaces when event occurs on canvas
            this.canvasElement.ownerDocument!.body.removeEventListener(k, this.canvasFocusCallback, { capture: false });
      });

      const w = this.canvasElement.ownerDocument!.defaultView;

      if (w) {
        w.removeEventListener('resize', this.onWindowResize);
      }

      this.canvasElement.removeEventListener('mousedown', this.onCanvasEvent);
      this.canvasElement.removeEventListener('mousemove', this.onCanvasEvent);
      this.canvasElement.ownerDocument!.removeEventListener('mouseup', this.onCanvasEvent);
    }

    initCanvas() {
      this.canvasElement.width = this.width;
      this.canvasElement.height = this.height;

      const w = this.canvasElement.ownerDocument!.defaultView;

      // w && (w.onresize = this.onWindowResize);
      if (w) {
        w.addEventListener('resize', this.onWindowResize);
      }

      this.onWindowResize();

      DrawingCanvasController.eventMap.forEach((_, k) => {
        if (k === 'touchend' || k === 'touchcancel') {
                this.canvasElement.ownerDocument!.addEventListener(k, this.onTouchEvent);
        } else {
          this.canvasElement.addEventListener(k, this.onTouchEvent);
        }

        // Prevent scrolling on touch interfaces when event occurs on canvas
        const { body } = this.canvasElement.ownerDocument!;

        body.addEventListener(k, this.canvasFocusCallback, { capture: false, passive: false });
      });

      this.canvasElement.addEventListener('mousedown', this.onCanvasEvent);
      this.canvasElement.addEventListener('mousemove', this.onCanvasEvent);
      // Setting listener to the document so lifting the mouse button anywhere will register as the tool being lifted
      this.canvasElement.ownerDocument!.addEventListener('mouseup', this.onCanvasEvent);
    }
}
