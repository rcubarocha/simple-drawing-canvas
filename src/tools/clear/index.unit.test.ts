import { clearMouseEventCallback, clearDrawingCallback, ClearTool } from '.';
import { CanvasConfig, BaseCanvasController, CanvasAction } from '../../canvas';

describe('clear mouse event callback', () => {
  let canvas: HTMLCanvasElement;
  let canvasConfig: CanvasConfig;
  let toolConfig: ClearTool;
  let mockController: BaseCanvasController;

  beforeEach(() => {
    canvas = document.createElement('canvas');
    canvas.width = 300;
    canvas.height = 300;
    
    canvas.getBoundingClientRect = jest.fn(() => ({
      x: 50,
      y: 50,
      width: 100,
      height: 100,
      top: 50,
      right: 150,
      bottom: 150,
      left: 50,
    } as DOMRect));

    canvasConfig = {
      scale: { x: 3, y: 3 },
      width: 300,
      height: 300,
      background: null
    };

    toolConfig = {
      style: "#333333"
    };

    mockController = {
      getCanvas: () => canvas,
      undo: (_) => {},
      redo: (_) => {},
      playDrawing: () => {},
      getDataURL: (t, q) => '',
      setBackground: (s) => {},
      teardown: () => {},
    }

  });

  // Mouse Down

  test('mouse down event should return correct result with Action Step', () => {
  
    const me = new window.MouseEvent('mousedown', {
      clientX: 100,
      clientY: 100,
    });
    
    const res = clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [] });

    const expected = {
      actionStatus: 'continue',
      actionUpdate: {
        actionStep: {
          tool: toolConfig,
          coords: {
            x: 150,
            y: 150,
          },
          state: 'down',
        },
        replacePrevStep: false,
      },
    };

    expect(res).toEqual(expected);
  });

  // Mouse Move

  test('mouse move event should always return result to ignore the event', () => {
  
    const me = new window.MouseEvent('mousemove', {
      clientX: 150,
      clientY: 150,
    });

    const expected = {
      actionStatus: 'continue',
    };
    
    expect (clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [] })).toEqual(expected);

    const downPrevStep = {
      tool: toolConfig,
      coords: {
        x: 150,
        y: 150,
      },
      state: 'up',
    };
    
    expect (clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [ downPrevStep ] })).toEqual(expected);
  });

  // Mouse Up

  test('mouse up event should throw if action history is in inconsistent state', () => {
  
    const me = new window.MouseEvent('mouseup', {
      clientX: 150,
      clientY: 150,
    });

    // Make sure the target of the event is the canvas
    canvas.dispatchEvent(me);

    const badPreviousStep = {
      tool: toolConfig,
      coords: {
        x: 150,
        y: 150,
      },
      state: 'up',
    };
    
    expect(() => {
      clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [ badPreviousStep ] })
    }).toThrow();
  });

  test('mouse up event should return result to cancel whole action if not targeting canvas', () => {
  
    const me = new window.MouseEvent('mouseup', {
      clientX: 150,
      clientY: 150,
    });

    const prevDown = {
      tool: toolConfig,
      coords: {
        x: 250,
        y: 250,
      },
      state: 'down',
    };

    const expected = {
      actionStatus: 'cancel',
    }
    
    expect(clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [ prevDown ] })).toEqual(expected);
  });

  test('mouse up event should return correct result with Action Step', () => {
  
    const me = new window.MouseEvent('mouseup', {
      clientX: 150,
      clientY: 150,
    });

    // Make sure the target of the event is the canvas
    canvas.dispatchEvent(me);

    const prevDown = {
      tool: toolConfig,
      coords: {
        x: 250,
        y: 250,
      },
      state: 'down',
    };
    
    const resPrevDown = clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [ prevDown ] });

    const expected = {
      actionStatus: 'end',
      actionUpdate: {
        actionStep: {
          tool: toolConfig,
          coords: {
            x: 300,
            y: 300,
          },
          state: 'up',
        },
        replacePrevStep: false,
      },
    };

    expect(resPrevDown).toEqual(expected);
  });

  // Other

  test('any other event should throw', () => {
  
    const me = new window.MouseEvent('mousewheel', {
      clientX: 150,
      clientY: 150,
    });
    
    expect(() => {
      clearMouseEventCallback.call(mockController, me, canvas, canvasConfig, toolConfig, { tool: 'clear', steps: [ ] });
    }).toThrow();
  });

});

// Drawing

describe('clear mouse action step callback', () => {
  let canvas: HTMLCanvasElement;
  let canvasConfig: CanvasConfig;
  let ctx: CanvasRenderingContext2D;
  let toolConfig: ClearTool;
  let emptyToolConfig: ClearTool;
  let actionHistory: CanvasAction<'clear', ClearTool>

  beforeEach(() => {
    canvas = document.createElement('canvas');
    canvas.width = 300;
    canvas.height = 300;
    
    canvas.getBoundingClientRect = jest.fn(() => ({
      x: 50,
      y: 50,
      width: 100,
      height: 100,
      top: 50,
      right: 150,
      bottom: 150,
      left: 50,
    } as DOMRect));

    ctx = canvas.getContext('2d')!;

    canvasConfig = {
      scale: { x: 3, y: 3 },
      width: 300,
      height: 300,
      background: null
    };

    toolConfig = {
      style: '#333333',
    };

    emptyToolConfig = {
    }

    actionHistory = {
      tool: 'clear',
      steps: [],
    }
  });

  test('the down step should not alter the canvas', () => {
  
    const downStep = {
      tool: toolConfig,
      coords: {
        x: 150,
        y: 150,
      },
      state: 'down',
    }

    clearDrawingCallback(downStep, actionHistory, canvas, canvasConfig);

    expect(canvas.toDataURL()).toMatchSnapshot()
  });

  test('the up step should result in an empty canvas when tool config is empty', () => {

    // Dirty up the canvas a bit
    ctx.fillStyle = "#666666";
    ctx.fillRect(50, 100, 200, 100);
  
    const downStep = {
      tool: emptyToolConfig,
      coords: {
        x: 150,
        y: 150,
      },
      state: 'down',
    };

    const upStep = {
      tool: emptyToolConfig,
      coords: {
        x: 50,
        y: 250,
      },
      state: 'up'
    };

    actionHistory.steps.push(downStep);

    clearDrawingCallback(upStep, actionHistory, canvas, canvasConfig);

    expect(canvas.toDataURL()).toMatchSnapshot()
  });

  test('the up step should result in canvas with solid fill according to config', () => {

    // Dirty up the canvas a bit
    ctx.fillStyle = "#666666";
    ctx.fillRect(50, 100, 200, 100);
  
    const downStep = {
      tool: toolConfig,
      coords: {
        x: 150,
        y: 150,
      },
      state: 'down',
    };

    const upStep = {
      tool: toolConfig,
      coords: {
        x: 50,
        y: 250,
      },
      state: 'up'
    };

    actionHistory.steps.push(downStep);

    clearDrawingCallback(upStep, actionHistory, canvas, canvasConfig);

    expect(canvas.toDataURL()).toMatchSnapshot()
  });
});