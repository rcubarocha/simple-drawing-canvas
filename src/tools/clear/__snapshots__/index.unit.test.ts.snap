// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clear mouse action step callback the down step should not alter the canvas 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AAAABmJLR0QA/wD/AP+gvaeTAAABdElEQVR4nO3BMQEAAADCoPVPbQwfoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD+Bn+3AAEkkD9cAAAAAElFTkSuQmCC"`;

exports[`clear mouse action step callback the up step should result in an empty canvas when tool config is empty 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AAAABmJLR0QA/wD/AP+gvaeTAAABdElEQVR4nO3BMQEAAADCoPVPbQwfoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD+Bn+3AAEkkD9cAAAAAElFTkSuQmCC"`;

exports[`clear mouse action step callback the up step should result in canvas with solid fill according to config 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AAAABmJLR0QA/wD/AP+gvaeTAAAD8ElEQVR4nO3UQQ0AIRDAwOMsrX9tYIEfaTKjoK+umdkfQMD/OgDglmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkGBaQYVhAhmEBGYYFZBgWkGFYQIZhARmGBWQYFpBhWECGYQEZhgVkHMVUA/BiZChoAAAAAElFTkSuQmCC"`;
